modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java

```diff
@@ -265,45 +378,116 @@ public MetricRegistry registry(String name) {
         return registries.computeIfAbsent(name, n -> {
             MetricRegistry mreg = new MetricRegistry(name, log);
 
-            notifyListeners(mreg, metricRegCreationLsnrs);
+            notifyListeners(mreg, metricRegCreationLsnrs, log);
 
             return mreg;
         });
     }
 
-    /** {@inheritDoc} */
-    @NotNull @Override public Iterator<MetricRegistry> iterator() {
-        return registries.values().iterator();
+    public <R extends MonitoringRow, D> void list(String name, String description,
+        Class<R> rowClazz, Supplier<ConcurrentMap<?, D>> data, Function<D, R> rowFunc, Consumer<D> clearer) {
+
+        Supplier<MonitoringList<R>> listCreator = () -> list(name, () -> new MonitoringListAdapter<>(name,
+            description,
+            rowClazz,
+            (MonitoringRowAttributeWalker<R>)walkers.get(rowClazz),
+            data.get(),
+            rowFunc));
+
+        //Create new instance of the list.
+        listCreator.get();
+
+        ctx.metric().addEnableListListener(listenOnlyEqual(name, identity(), n -> listCreator.get()));
```

Comment:
A very complicated logic here. It's up to 5-6 nested consumers/suppliers used for event listener. Perhaps it can be simplified somehow (but I'm not sure). In current implementation is hard to understand such a code.

Reply:
1. We create a list which exports map content.
2. On list removal event(via `IgniteMBean#disableList`) `rowClearer` will clear data needed only for list(if applicable).
3. On list enable event(via `IgniteMBean#enableList`) it will be recreated(fresh instance of the `data` map will be obtained.

I think it pretty simple and easy to use approach.

Reply:
And actually it's not a listener. It's something like "list enabler", it's lack of abstraction here since listener should only react to this event, but not performing event itself.

Reply:
I doesn't understand your proposal, actually. 
What is the different between "enable list event" and "list enabler"?

Reply:
I don't say that I didn't understand the code, I say that code is hard to understand because it uses 5-6 nested lambdas. It's certainly not a pretty simple.  

I don't propose to rename it to "list enabler", I say that "listener" is a bad abstraction for this case.

Perhaps code can be simplified in this way:
1. Create maps for list creators and removers (name -> closure) instead of lists for listeners 
2. On enabling/disabling just call closure from this map by name.
3. Get rid of as most nested lambdas as possible. Now it looks redundant. Ideally, for list creation case there should be only two labdas: one for list creation closure, and one for `computeIfAbsent` mapping function.